File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {

        List<Integer> ret = new ArrayList<>();
        int product = 1;
        int zeroCount = 0;

        
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) zeroCount++;
            else product *= nums[i];
        }

        
        for (int i = 0; i < nums.length; i++) {
            if (zeroCount > 1) {
                ret.add(0);
            }
            else if (zeroCount == 1) {
                ret.add(nums[i] == 0 ? product : 0);
            }
            else {
                ret.add(product / nums[i]);
            }
        }

        
        int[] result = new int[ret.size()];
        for (int i = 0; i < ret.size(); i++) {
            result[i] = ret.get(i);
        }

        return result;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
    // let ret = [];
    // let values = [];
    // for (let i = 0; i < nums.length; i++) {
    //     for (let j = 0; j < nums.length; j++) {
    //         if (i == j) { continue; }
    //         values.push(nums[j]);
    //     }
    //     ret.push(values.reduce((a, b) => a * b, 1));
    //     values=[];
    // }
    // return ret;

    let ret = [];
    let product = 1;
    let zeroCount = 0;

   
    for (let num of nums) {
        if (num == 0) zeroCount++;
        else product *= num;
    }

    for (let num of nums) {
        if (zeroCount > 1) {
            ret.push(0);
        }
        else if (zeroCount == 1) {
            ret.push(num == 0 ? product : 0);
        }
        else {
            ret.push(product / num);
        }
    }

    return ret;

};